### PR TITLE
[stable10] allow reenabling users on sync

### PR DESCRIPTION
**Pull request by butonic**
- Allow re-enabling users on `occ user:sync`,
- added email and display name to the output to better identify users
- updated the help message

```console
% ./occ user:sync "OCA\User_LDAP\User_Proxy" --help       
Usage:
  user:sync [options] [--] [<backend-class>]

Arguments:
  backend-class                                        The PHP class name - e.g., "OCA\User_LDAP\User_Proxy". Please wrap the class name in double quotes. You can use the option --list to list all known backend classes.

Options:
  -l, --list                                           List all known backend classes
  -u, --uid=UID                                        sync only the user with the given user id
  -s, --seenOnly                                       sync only seen users
  -c, --showCount                                      calculate user count before syncing
  -m, --missing-account-action=MISSING-ACCOUNT-ACTION  Action to take if the account isn't connected to a backend any longer. Options are "disable" and "remove". Note that removing the account will also remove the stored data and files for that account.
  -r, --re-enable                                      When syncing multiple accounts re-enable accounts that are disabled in ownCloud but available in the synced backend.
  -h, --help                                           Display this help message
  -q, --quiet                                          Do not output any message
  -V, --version                                        Display this application version
      --ansi                                           Force ANSI output
      --no-ansi                                        Disable ANSI output
  -n, --no-interaction                                 Do not ask any interactive question
      --no-warnings                                    Skip global warnings, show command output only
  -v|vv|vvv, --verbose                                 Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug

Help:
  Synchronize users from a given backend to the accounts table.
% ./occ user:sync "OCA\User_LDAP\User_Proxy" -m disable -r
Analysing all users ...
    6 [============================]

No removed users have been detected.

No existing accounts to re-enable.

Insert new and update existing users ...
    4 [============================]

% ./occ user:sync "OCA\User_LDAP\User_Proxy" -m disable -r
Analysing all users ...
    6 [============================]

Following users are no longer known with the connected backend.
Disabling accounts:
9F625F70-08DD-4838-AD52-7DE1F72DBE30, Bobbie, bobbie@example.org disabled
53CDB5AC-B02E-4A49-8FEF-001A13725777, David, dave@example.org disabled
34C3F461-90FE-417C-ADC5-CE97FE5B8E72, Carol, carol@example.org disabled

No existing accounts to re-enable.

Insert new and update existing users ...
    1 [============================]

% ./occ user:sync "OCA\User_LDAP\User_Proxy" -m disable -r
Analysing all users ...
    6 [============================]

Following users are no longer known with the connected backend.
Disabling accounts:
53CDB5AC-B02E-4A49-8FEF-001A13725777, David, dave@example.org skipped, already disabled
34C3F461-90FE-417C-ADC5-CE97FE5B8E72, Carol, carol@example.org skipped, already disabled
B5275C13-6466-43FD-A129-A12A6D3D9A0D, Alicia3, alicia3@example.org disabled

Re-enabling accounts:
9F625F70-08DD-4838-AD52-7DE1F72DBE30, Bobbie, bobbie@example.org enabled

Insert new and update existing users ...
    1 [============================]

% ./occ user:sync "OCA\User_LDAP\User_Proxy" -m disable -r
Analysing all users ...
    6 [============================]

No removed users have been detected.

Re-enabling accounts:
53CDB5AC-B02E-4A49-8FEF-001A13725777, David, dave@example.org enabled
34C3F461-90FE-417C-ADC5-CE97FE5B8E72, Carol, carol@example.org enabled
B5275C13-6466-43FD-A129-A12A6D3D9A0D, Alicia3, alicia3@example.org enabled

Insert new and update existing users ...
    4 [============================]

```

### CODE REVIEW

**Comment by butonic** [comment]
- [ ] forward port to master

File: core/Command/User/SyncBackend.php [hunk 1]
```diff
@@ -265,95 +283,131 @@ private function getBackend($backend) {
...
-								});
+							$output->writeln('Disabling the accounts');
+							$this->doActionForAccountUids(
+								$removedUsers,
+								$enableAction,
+								$writeNotExisting
+							);
 							break;
 						case 'remove':
 							$output->writeln('Proceeding to remove the accounts');
```
Comment (phil-davis) [inline core/Command/User/SyncBackend.php]: In code above, you changed this to:
```
$output->writeln('Deleting accounts:');
```
so could be changed here also to be consistent.

File: core/Command/User/SyncBackend.php [hunk 2]
```diff
@@ -219,28 +231,34 @@ private function syncMultipleUsers (
...
 		$output->writeln("Syncing $uid ...");
 		$users = $backend->getUsers($uid, 2);
 		if (count($users) > 1) {
-			$output->writeln("Multiple users returned from backend for input: $uid. Cancelling sync.");
-			return 1;
-		} elseif (count($users) === 1) {
+			throw new \LengthException("Multiple users returned from backend for input: $uid. Cancelling sync.");
+		}
+
+		$dummy = new Account(); // to prevent null pointer wen writing messages
```
Comment (phil-davis) [inline core/Command/User/SyncBackend.php]: s/wen/when

**Comment by butonic** [comment]
streamlined the output messages

**Comment by tomneedham** [comment]
Works nicely here with master ldap. Just checking out the code...

**Comment by tomneedham** [comment]
restarted drone as it had a hiccup getting to github on one of the jobs

File: core/Command/User/SyncBackend.php [hunk 3]
```diff
@@ -259,101 +286,136 @@ private function getBackend($backend) {
 	}
 
 	/**
-	 * @param array $uids a list of uids to the the action
+	 * @param array $uidToAccountMap a list of uids to the the action
```
Comment (tomneedham) [inline core/Command/User/SyncBackend.php]: s/action/account ? This is a map of uid strings to account objects right?

File: core/Command/User/SyncBackend.php [hunk 4]
```diff
@@ -259,101 +286,136 @@ private function getBackend($backend) {
...
 	 * @param callable $callbackExists the callback used if the account for the uid exists. The
 	 * uid and the specific account will be passed as parameter to the callback in that order
 	 * @param callable $callbackMissing the callback used if the account doesn't exists. The uid (not
 	 * the account) will be passed as parameter to the callback
```
Comment (tomneedham) [inline core/Command/User/SyncBackend.php]: seems it is called with the uid and the account as params from line 299

**Comment by tomneedham** [comment]
pushed some very basic tests that increase coverage by at least executing the missing methods and checking they dont explode with some basic input. `analyzeExistingUsers` is quite hard to test because of the callback nature of the method - not sure how we can improve this?

**Comment by tomneedham** [comment]
also fixed some comment hiccups

**Comment by PVince81** [comment]
conflict

**Review by tomneedham — APPROVED** [review]

**Comment by PVince81** [comment]
@tomneedham @butonic please port to master

**Comment by PVince81** [comment]
master: https://github.com/owncloud/core/pull/31174

**Comment by mmattel** [comment]
@butonic 
Because of just writing the docs for that,

the text: on the option ``--list`` ``List all known backend classes``
should´nt it be better ``List all enabled backend classes`` (versus known)?
Because if I type on my dev instance which has no other than local DB,
``sudo -uwww-data ./occ user:sync --list`` I only get ``OC\User\Database``

and
```
  -u, --uid=UID           sync only the user with the given user id
  -s, --seenOnly          sync only seen users
  -c, --showCount         calculate user count before syncing
```
The description text should start with capital letters as all the others do

**Comment by settermjd** [comment]
I'm late to this, but `--seenOnly` might be just that much more intuitive when written as `--onlySeen`, perhaps.

**Comment by mmattel** [comment]
No worry,
not too late, create a PR changing this and I will reflect it in the doc immediately 😄 
But what means "seen"... ? (yes I can translate it, but "seen" in the context...)

**Pull request merged**